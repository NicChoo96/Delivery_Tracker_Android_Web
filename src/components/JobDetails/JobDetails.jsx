import React, { useState, useEffect, useCallback } from 'react';
import {
  Grid,
  Card,
  CardContent,
} from '@material-ui/core';
import './JobDetails.scss';

import { Contractor, MapDisplay } from '../index';
import {
  timestampConversion,
  approximateAddress,
  immutableSort,
  flatternObjectToArray,
} from '../../utils';

/**
 * This component will render the job details
 * @category UI Component
 * @subcategory JobDetails
 * @class JobDetails
 * @param {number} startTime Timestamp when the job starts
 * @param {number} endTime Timestamp when the job ends
 * @param {number} phoneNo Phone number of contractor
 * @param {string} name Name of Contractor
 * @param {Array<{locationData}>} locationHistory array of all job history information
 * @return {ReactComponent}
 */
function JobDetails({ startTime = 0, endTime = 0, phoneNo, name, company, locationHistory }) {
  const labelSize = 4;
  const textSize = 8;

  const [parsedLocations, setParsedLocations] = useState({});
  const [displayLocation, setDisplayLocation] = useState([]);

  const memoizedApproximatAddress = useCallback(approximateAddress, []);

  // fetches approximate addresses from google
  // then asynchronously update the state of parsedLocation
  useEffect(() => {
    // iterating through the entire locationHistory prop
    // to get every lat long and write to the parseLocataions
    // state individually for each location
    if (locationHistory && locationHistory.length && locationHistory.length > 0) {
      locationHistory.filter(({ lat, lng, dateTime }) => {
        memoizedApproximatAddress(lat, lng)
          .then(response => {
            setParsedLocations((state) => {
              const newState = { ...state };
              newState[dateTime] = { approximateLocation: response, dateTime: timestampConversion(dateTime) };
              return newState;
            });
          });
          return false;
      });
    }
  }, [memoizedApproximatAddress, locationHistory]);

  // fetches approximate addresses from google
  // convert the parsedLocation object into a sorted
  // displayable array with descending timestamp
  useEffect(() => {
    const locationEntries = flatternObjectToArray(parsedLocations, true).map(data => {
      return { timeStamp: data.__objectKey__, approximateLocation: data.approximateLocation, dateTime: data.dateTime };
    })
    setDisplayLocation(immutableSort(locationEntries, (a, b) => {
      return b.timeStamp - a.timeStamp;
    }));
  }, [parsedLocations]);

  return (
    <>
      <Card>
        <CardContent>
          <Grid container direction="row" spacing={5} justify="center">
            <Grid item md={7} sm={5} container>

              <Grid item xs={labelSize}>
                <h3 className="job-details__label">Last Location:</h3>
              </Grid>
              <Grid item xs={textSize}>
                <p
                  data-testid="job-details__value-location"
                  className="app__text-cursor"
                >
                  {(displayLocation.length > 0) ? displayLocation[displayLocation.length - 1].approximateLocation : 'No location'}
                </p>
              </Grid>

              <Grid item xs={labelSize}>
                <h3 className="job-details__label">Start Time:</h3>
              </Grid>
              <Grid item xs={textSize}>
                <p
                  data-testid="job-details__value-starttime"
                  className="app__text-cursor"
                >
                  {timestampConversion(startTime, true, true)}
                </p>
              </Grid>

              <Grid item xs={labelSize}>
                <h3 className="job-details__label">End Time:</h3>
              </Grid>
              <Grid item xs={textSize}>
                <p
                  data-testid="job-details__value-endtime"
                  className="app__text-cursor"
                >
                  {endTime ? timestampConversion(endTime, true, true) : 'Job In Progress'}
                </p>
              </Grid>
              <Grid item xs={textSize + labelSize}>
                <h3 className="job-details__label">Contractor Information:</h3>
                <Contractor companyName={company} phoneNumber={phoneNo} contractorName={name} />
              </Grid>
            </Grid>

            <Grid item md={5} sm={7}>
              <Card className="job-details__full-length-height">
                <div className="job-details__mapholder">

                  {
                    (!(locationHistory && locationHistory.length && locationHistory.length > 0))
                      ? <p className="job-location__no-data-message app__no-data-message">No Location Data Available</p>
                      : (
                        <MapDisplay places={locationHistory} />
                      )
                  }
                </div>
              </Card>
            </Grid>
          </Grid>
        </CardContent>
      </Card>
      <br />
    </>
  );
}

export default JobDetails;
