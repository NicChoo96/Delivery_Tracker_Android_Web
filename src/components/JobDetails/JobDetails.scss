.job-details__mapholder {
  position: relative;
  width: auto;
  height: 100%;
}

.job-details__full-length-height {
  height: 100%;
}

/* ========== Job Location Report ========== */

.job-location__data-control-group {
  user-select: none;
}

.job-location__no-data-message {
  height: 5em !important;
}

