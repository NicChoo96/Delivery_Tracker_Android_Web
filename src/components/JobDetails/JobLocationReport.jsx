import React, { useState, useEffect, useCallback } from 'react';
import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  FormGroup,
  FormControlLabel,
  Switch,
} from '@material-ui/core';

import {
  timestampConversion,
  approximateAddress,
  immutableSort,
  flatternObjectToArray,
} from '../../utils';

/**
 * Renders a list of all the locations the contractors has been to and the accompanying time
 * @category UI Component
 * @subcategory JobDetails
 * @class JobLocationReport
 * @param {Array<{locationData}>} locationHistory array of all job history information
 * @return {ReactComponent}
 */
function JobLocationReport({ locationHistory }) {
  const [parsedLocations, setParsedLocations] = useState({});
  const [preSortedLocations, setPreSortedLocations] = useState([]);
  const [displayLocations, setDisplayLocations] = useState([]);
  const [sortDescending, setSortDescending] = useState(true);

  const memoizedApproximatAddress = useCallback(approximateAddress, []);

  const descendingHandler = (event) => {
    setSortDescending(event.target.checked);
  }

  useEffect(() => {
    locationHistory.filter(({ lat, lng, dateTime }) => {
      memoizedApproximatAddress(lat, lng)
        .then(response => {
          setParsedLocations((state) => {
            const newState = { ...state };
            newState[dateTime] = { approximateLocation: response, dateTime: timestampConversion(dateTime, true, true) };
            return newState;
          });
        });
      return false;
    });
  }, [memoizedApproximatAddress, locationHistory]);

  useEffect(() => {
    const locationEntries = flatternObjectToArray(parsedLocations, true).map(data => {
      return { timeStamp: data.__objectKey__, approximateLocation: data.approximateLocation, dateTime: data.dateTime };
    })
    setPreSortedLocations(locationEntries);
  }, [parsedLocations]);

  useEffect(() => {
    setDisplayLocations(immutableSort(preSortedLocations, (a, b) => {
      return (b.timeStamp - a.timeStamp) * ((sortDescending) ? 1 : -1);
    }));

    return () => setDisplayLocations([]);
  }, [preSortedLocations, sortDescending]);

  return (
    <>
      <h2>Location Report</h2>

      <FormGroup row className="job-location__data-control-group">
        <FormControlLabel
          disabled={!(displayLocations && displayLocations.length > 1)}
          control={
            <Switch
              checked={sortDescending}
              name="Show Descending"
              onChange={descendingHandler}
              color="primary"
            />}
          label="Date Time Descending"
        />
      </FormGroup>
      <TableContainer component={Paper} className="app-table">
        <Table aria-label="a location history table" size="small" >
          <TableHead>
            <TableRow>
              <TableCell align="left">
                <span className="app-table__header-cell">
                  Date Time Started
                  </span>
              </TableCell>
              <TableCell align="left">
                <span className="app-table__header-cell">
                  Location
                  </span>
              </TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {(displayLocations.length > 0) ? displayLocations.map(data => (
              <TableRow key={data.timeStamp} className="app-table__row">
                <TableCell component="th" scope="data">
                  <span className="app__text-cursor">
                    {data.dateTime}
                  </span>
                </TableCell>
                <TableCell align="left">
                  <span className="app__text-cursor">
                    {data.approximateLocation}
                  </span>
                </TableCell>
              </TableRow>
            ))
              : <TableRow key={69}>
                <TableCell align="center" colSpan={8}>
                  <span className="app__no-data-message job-location__no-data-message">No data available.</span>
                </TableCell>
              </TableRow>
            }
          </TableBody>
        </Table>
      </TableContainer>
    </>
  );
}

export default JobLocationReport;
