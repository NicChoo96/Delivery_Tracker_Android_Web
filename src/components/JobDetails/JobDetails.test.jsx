import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';
import { cleanup, render, wait } from '@testing-library/react';
import renderer from 'react-test-renderer';
import JobDetails from '../../components/JobDetails/JobDetails';

afterEach(cleanup);

describe('JobDetails', () => {
  const jobDetailsTestProps = {
    locationHistory: [],
    startTime: '1602939179290',
    endTime: '1602939179400',
    phoneNo: '69696969',
    name: 'I am a real person',
    company: 'This is a real company',
  };

  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<JobDetails />, div);
  });

  test('render value', async () => {
    const { getByTestId } = render(<JobDetails {...jobDetailsTestProps} />);

    await wait();
    expect(getByTestId('job-details__value-starttime').innerHTML).toBe(
      '17 Oct 2020 - 20:52:59 PM'
    );
    expect(getByTestId('job-details__value-endtime').innerHTML).toBe(
      '17 Oct 2020 - 20:52:59 PM'
    );
  });

  test('matches snapshot', () => {
    const tree = renderer
      .create(<JobDetails {...jobDetailsTestProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
