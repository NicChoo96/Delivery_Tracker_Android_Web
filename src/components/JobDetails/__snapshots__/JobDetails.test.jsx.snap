// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobDetails matches snapshot 1`] = `
Array [
  <div
    className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiCardContent-root"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5 MuiGrid-justify-xs-center"
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-sm-5 MuiGrid-grid-md-7"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <h3
              className="job-details__label"
            >
              Last Location:
            </h3>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
          >
            <p
              className="app__text-cursor"
              data-testid="job-details__value-location"
            >
              No location
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <h3
              className="job-details__label"
            >
              Start Time:
            </h3>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
          >
            <p
              className="app__text-cursor"
              data-testid="job-details__value-starttime"
            >
              17 Oct 2020 - 20:52:59 PM
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <h3
              className="job-details__label"
            >
              End Time:
            </h3>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
          >
            <p
              className="app__text-cursor"
              data-testid="job-details__value-endtime"
            >
              17 Oct 2020 - 20:52:59 PM
            </p>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <h3
              className="job-details__label"
            >
              Contractor Information:
            </h3>
            <div
              className="contractor__card-width"
            >
              <div
                className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
              >
                <div
                  className="MuiCardContent-root contractor__card-wrapper"
                >
                  <h1
                    className="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextPrimary"
                  >
                    <span
                      className="contractor_name"
                    >
                      I am a real person
                    </span>
                  </h1>
                  <p
                    className="MuiTypography-root MuiTypography-caption"
                  >
                    from 
                    <span
                      className="contractor__company-name"
                    >
                      This is a real company
                    </span>
                  </p>
                  <p
                    className="MuiTypography-root MuiTypography-body2"
                  >
                    Mobile number: 
                    69696969
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-7 MuiGrid-grid-md-5"
        >
          <div
            className="MuiPaper-root MuiCard-root job-details__full-length-height MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              className="job-details__mapholder"
            >
              <p
                className="job-location__no-data-message app__no-data-message"
              >
                No Location Data Available
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <br />,
]
`;
